File: bitcache/three_stage.py
"""Three-stage progressive retrieval: binary → 4-bit → float32.

Stage 1: Binary Hamming scan (1-bit per dim). Cheap. Selects broad candidates.
Stage 2: 4-bit quantized inner product on candidates. Medium cost. Refines ranking.
Stage 3: Float32 inner product on top candidates. Precise. Final ranking.

Memory hierarchy:
  Stage 1 codes: 96 bytes/vector (d=768)  — always in RAM
  Stage 2 codes: 384 bytes/vector (d=768) — in RAM
  Stage 3 vectors: 3072 bytes/vector (d=768) — loaded on demand
"""

import numpy as np
from typing import Optional

from bitcache.quantize import quantize
from bitcache.search import hamming_distance_single


class ThreeStageIndex:
    """Three-stage retrieval: binary filter → 4-bit rerank → float32 rerank.

    Each stage narrows the candidate set with increasing precision.
    Achieves higher recall than two-stage by adding an intermediate
    quantized scoring step that catches candidates missed by binary
    ranking alone.

    Args:
        dim: Vector dimensionality.
        stage1_factor: Candidates from Stage 1 = k * stage1_factor.
        stage2_factor: Candidates from Stage 2 = k * stage2_factor.
        n_bits: Quantization bits for Stage 2 (default 4).
    """

    def __init__(
        self,
        dim: int,
        stage1_factor: int = 200,
        stage2_factor: int = 20,
        n_bits: int = 4,
    ):
        if dim <= 0:
            raise ValueError("dim must be positive")
        self.dim = dim
        self.stage1_factor = stage1_factor
        self.stage2_factor = stage2_factor
        self.n_bits = n_bits
        self.n_levels = 2 ** n_bits  # 16 for 4-bit

        self._binary_codes: Optional[np.ndarray] = None
        self._quant_codes: Optional[np.ndarray] = None
        self._vectors: Optional[np.ndarray] = None
        self._centroids: Optional[np.ndarray] = None
        self._n_vectors: int = 0

    def add(self, vectors: np.ndarray) -> None:
        """Add vectors to the index.

        Args:
            vectors: Float array of shape (n, dim).
        """
        vectors = np.asarray(vectors, dtype=np.float32)
        if vectors.ndim == 1:
            vectors = vectors[None, :]
        if vectors.shape[1] != self.dim:
            raise ValueError(f"expected dim={self.dim}, got {vectors.shape[1]}")

        # Normalize
        norms = np.linalg.norm(vectors, axis=1, keepdims=True)
        norms = np.where(norms > 1e-10, norms, 1.0)
        unit_vectors = vectors / norms

        # Stage 1: binary codes
        binary_codes = quantize(unit_vectors)

        # Stage 2: 4-bit uniform scalar quantization per dimension
        quant_codes, centroids = self._quantize_4bit(unit_vectors)

        if self._binary_codes is None:
            self._binary_codes = binary_codes
            self._quant_codes = quant_codes
            self._vectors = unit_vectors
            self._centroids = centroids
        else:
            self._binary_codes = np.vstack([self._binary_codes, binary_codes])
            self._quant_codes = np.vstack([self._quant_codes, quant_codes])
            self._vectors = np.vstack([self._vectors, unit_vectors])
            # Recompute centroids on full data
            self._quant_codes, self._centroids = self._quantize_4bit(self._vectors)

        self._n_vectors = len(self._binary_codes)

    def search(self, query: np.ndarray, k: int = 10) -> tuple[np.ndarray, np.ndarray]:
        """Three-stage search.

        Args:
            query: Float vector of shape (dim,).
            k: Number of final results.

        Returns:
            Tuple of (scores, indices). Scores are float32 inner products.
        """
        if self._binary_codes is None or self._n_vectors == 0:
            return np.array([], dtype=np.float32), np.array([], dtype=np.int64)

        query = np.asarray(query, dtype=np.float32).ravel()
        norm = np.linalg.norm(query)
        query_unit = query / norm if norm > 1e-10 else query

        # Stage 1: binary Hamming — broad candidate selection
        query_binary = quantize(query_unit[None, :])[0]
        n_stage1 = min(k * self.stage1_factor, self._n_vectors)

        hamming_dists = hamming_distance_single(query_binary, self._binary_codes)

        if n_stage1 < self._n_vectors:
            stage1_indices = np.argpartition(hamming_dists, n_stage1)[:n_stage1]
        else:
            stage1_indices = np.arange(self._n_vectors)

        # Stage 2: 4-bit quantized inner product — refine candidates
        n_stage2 = min(k * self.stage2_factor, len(stage1_indices))

        query_quant_scores = self._score_quantized(query_unit, stage1_indices)

        if n_stage2 < len(stage1_indices):
            stage2_local = np.argpartition(query_quant_scores, -n_stage2)[-n_stage2:]
        else:
            stage2_local = np.arange(len(stage1_indices))

        stage2_indices = stage1_indices[stage2_local]

        # Stage 3: float32 inner product — precise final ranking
        candidate_vectors = self._vectors[stage2_indices]
        scores = candidate_vectors @ query_unit

        top_k_local = np.argsort(scores)[::-1][:k]
        final_indices = stage2_indices[top_k_local]
        final_scores = scores[top_k_local]

        return final_scores, final_indices

    def search_batch(self, queries: np.ndarray, k: int = 10) -> tuple[np.ndarray, np.ndarray]:
        """Batch three-stage search.

        Args:
            queries: Float array of shape (nq, dim).
            k: Number of results per query.

        Returns:
            Tuple of (scores, indices) each shape (nq, k).
        """
        queries = np.asarray(queries, dtype=np.float32)
        if queries.ndim == 1:
            queries = queries[None, :]

        nq = len(queries)
        all_scores = np.zeros((nq, k), dtype=np.float32)
        all_indices = np.zeros((nq, k), dtype=np.int64)

        for i in range(nq):
            scores, indices = self.search(queries[i], k=k)
            n = len(scores)
            all_scores[i, :n] = scores
            all_indices[i, :n] = indices

        return all_scores, all_indices

    def _quantize_4bit(self, vectors: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
        """Uniform scalar quantization to n_bits per dimension.

        Maps each coordinate to one of 2^n_bits levels based on
        the observed min/max range per dimension.

        Returns:
            Tuple of (codes as uint8 array shape (n, dim), centroids shape (n_levels, dim)).
        """
        # Compute per-dimension min/max
        vmin = vectors.min(axis=0)
        vmax = vectors.max(axis=0)
        vrange = vmax - vmin
        vrange = np.where(vrange > 1e-10, vrange, 1.0)

        # Quantize to [0, n_levels-1]
        normalized = (vectors - vmin) / vrange
        codes = np.clip(normalized * self.n_levels, 0, self.n_levels - 1).astype(np.uint8)

        # Compute centroids (midpoint of each bucket)
        centroids = np.zeros((self.n_levels, self.dim), dtype=np.float32)
        for level in range(self.n_levels):
            centroids[level] = vmin + (level + 0.5) / self.n_levels * vrange

        # Store min/range for query quantization
        self._vmin = vmin
        self._vrange = vrange

        return codes, centroids

    def _score_quantized(self, query_unit: np.ndarray, indices: np.ndarray) -> np.ndarray:
        """Compute approximate inner product using 4-bit codes.

        For each candidate, reconstruct the vector from its quantized codes
        using the centroid lookup table, then dot with query.
        """
        # Precompute query dot with each centroid per dimension
        # query_lut[level, dim] = query[dim] * centroid[level, dim]
        # Score for vector i = sum over dims of query_lut[code[i, dim], dim]

        # Build lookup table: for each dimension, score contribution per level
        lut = np.zeros((self.n_levels, self.dim), dtype=np.float32)
        for level in range(self.n_levels):
            lut[level] = query_unit * self._centroids[level]

        # Score each candidate
        candidate_codes = self._quant_codes[indices]  # shape (m, dim)
        n_candidates = len(indices)

        scores = np.zeros(n_candidates, dtype=np.float32)
        for i in range(n_candidates):
            # Sum lut[code[i,j], j] for all j
            codes_i = candidate_codes[i]
            scores[i] = sum(lut[codes_i[j], j] for j in range(self.dim))

        return scores

    def __len__(self) -> int:
        return self._n_vectors

    @property
    def memory_usage(self) -> dict:
        """Memory breakdown by stage."""
        if self._n_vectors == 0:
            return {"stage1": 0, "stage2": 0, "stage3": 0, "total": 0}
        s1 = self._binary_codes.nbytes
        s2 = self._quant_codes.nbytes
        s3 = self._vectors.nbytes
        return {
            "stage1_bytes": s1,
            "stage2_bytes": s2,
            "stage3_bytes": s3,
            "total_bytes": s1 + s2 + s3,
            "stage1_mb": round(s1 / 1024**2, 1),
            "stage2_mb": round(s2 / 1024**2, 1),
            "stage3_mb": round(s3 / 1024**2, 1),
        }

File: src/three_stage.rs
//! Three-stage progressive retrieval: binary → 4-bit → float32.
//!
//! Stage 1: Binary Hamming scan (1-bit per dim). Cheap. Selects broad candidates.
//! Stage 2: 4-bit quantized inner product on candidates. Medium cost. Refines ranking.
//! Stage 3: Float32 inner product on top candidates. Precise. Final ranking.

use crate::quantize::{normalize, normalize_batch, quantize, quantize_batch};
use crate::search::{hamming_distance_one_to_many, inner_product};

/// Three-stage retrieval: binary filter → 4-bit rerank → float32 rerank.
pub struct ThreeStageIndex {
    dim: usize,
    n_bytes: usize,
    n_vectors: usize,
    stage1_factor: usize,
    stage2_factor: usize,
    n_bits: usize,
    n_levels: usize,
    binary_codes: Vec<u8>,   // n * n_bytes
    quant_codes: Vec<u8>,    // n * dim (4-bit stored as u8 per dim)
    vectors: Vec<f32>,       // n * dim
    centroids: Vec<f32>,     // n_levels * dim
    vmin: Vec<f32>,          // dim
    vrange: Vec<f32>,        // dim
}

impl ThreeStageIndex {
    pub fn new(dim: usize, stage1_factor: usize, stage2_factor: usize, n_bits: usize) -> Self {
        assert!(dim > 0);
        assert!((2..=8).contains(&n_bits));
        Self {
            dim,
            n_bytes: (dim + 7) / 8,
            n_vectors: 0,
            stage1_factor,
            stage2_factor,
            n_bits,
            n_levels: 1 << n_bits,
            binary_codes: Vec::new(),
            quant_codes: Vec::new(),
            vectors: Vec::new(),
            centroids: Vec::new(),
            vmin: vec![0.0; dim],
            vrange: vec![1.0; dim],
        }
    }

    /// Add vectors to the index.
    pub fn add(&mut self, vectors: &[f32]) {
        let n = vectors.len() / self.dim;
        assert_eq!(vectors.len(), n * self.dim);

        let mut unit_vectors = vectors.to_vec();
        normalize_batch(&mut unit_vectors, n, self.dim);

        // Stage 1: binary codes
        let binary_codes = quantize_batch(&unit_vectors, n, self.dim);

        // Compute per-dimension min/max for scalar quantization
        let mut vmin = vec![f32::MAX; self.dim];
        let mut vmax = vec![f32::MIN; self.dim];
        for i in 0..n {
            for d in 0..self.dim {
                let val = unit_vectors[i * self.dim + d];
                if val < vmin[d] { vmin[d] = val; }
                if val > vmax[d] { vmax[d] = val; }
            }
        }
        // Include existing vectors in range computation
        for i in 0..self.n_vectors {
            for d in 0..self.dim {
                let val = self.vectors[i * self.dim + d];
                if val < vmin[d] { vmin[d] = val; }
                if val > vmax[d] { vmax[d] = val; }
            }
        }

        let vrange: Vec<f32> = vmin.iter().zip(vmax.iter())
            .map(|(mn, mx)| {
                let r = mx - mn;
                if r > 1e-10 { r } else { 1.0 }
            })
            .collect();

        self.vmin = vmin;
        self.vrange = vrange;

        // Stage 2: scalar quantize all vectors (including existing)
        let total_n = self.n_vectors + n;
        let mut all_vectors = self.vectors.clone();
        all_vectors.extend_from_slice(&unit_vectors);

        let quant_codes = self.scalar_quantize(&all_vectors, total_n);

        // Compute centroids
        let centroids = self.compute_centroids();

        self.binary_codes.extend_from_slice(&binary_codes);
        self.vectors.extend_from_slice(&unit_vectors);
        self.quant_codes = quant_codes;
        self.centroids = centroids;
        self.n_vectors = total_n;
    }

    /// Three-stage search.
    pub fn search(&self, query: &[f32], k: usize) -> (Vec<f32>, Vec<usize>) {
        if self.n_vectors == 0 {
            return (vec![], vec![]);
        }
        assert_eq!(query.len(), self.dim);

        let mut query_unit = query.to_vec();
        normalize(&mut query_unit);

        // Stage 1: binary Hamming
        let query_binary = quantize(&query_unit);
        let n_stage1 = (k * self.stage1_factor).min(self.n_vectors);
        let hamming_dists = hamming_distance_one_to_many(&query_binary, &self.binary_codes, self.n_vectors, self.n_bytes);

        let mut stage1: Vec<(u32, usize)> = hamming_dists.into_iter().enumerate().map(|(i, d)| (d, i)).collect();
        if n_stage1 < stage1.len() {
            stage1.select_nth_unstable_by(n_stage1 - 1, |a, b| a.0.cmp(&b.0));
            stage1.truncate(n_stage1);
        }
        let stage1_indices: Vec<usize> = stage1.iter().map(|(_, i)| *i).collect();

        // Stage 2: quantized inner product
        let n_stage2 = (k * self.stage2_factor).min(stage1_indices.len());
        let quant_scores = self.score_quantized(&query_unit, &stage1_indices);

        let mut stage2: Vec<(f32, usize)> = quant_scores.into_iter()
            .zip(stage1_indices.iter())
            .map(|(s, &i)| (s, i))
            .collect();
        stage2.sort_unstable_by(|a, b| b.0.partial_cmp(&a.0).unwrap());
        stage2.truncate(n_stage2);

        // Stage 3: float32 rerank
        let mut final_scored: Vec<(f32, usize)> = stage2
            .iter()
            .map(|&(_, idx)| {
                let v = &self.vectors[idx * self.dim..(idx + 1) * self.dim];
                let score = inner_product(&query_unit, v);
                (score, idx)
            })
            .collect();

        final_scored.sort_unstable_by(|a, b| b.0.partial_cmp(&a.0).unwrap());
        final_scored.truncate(k);

        let scores: Vec<f32> = final_scored.iter().map(|(s, _)| *s).collect();
        let indices: Vec<usize> = final_scored.iter().map(|(_, i)| *i).collect();
        (scores, indices)
    }

    fn scalar_quantize(&self, vectors: &[f32], n: usize) -> Vec<u8> {
        let mut codes = vec![0u8; n * self.dim];
        for i in 0..n {
            for d in 0..self.dim {
                let val = vectors[i * self.dim + d];
                let normalized = (val - self.vmin[d]) / self.vrange[d];
                let level = (normalized * self.n_levels as f32).clamp(0.0, (self.n_levels - 1) as f32) as u8;
                codes[i * self.dim + d] = level;
            }
        }
        codes
    }

    fn compute_centroids(&self) -> Vec<f32> {
        let mut centroids = vec![0.0f32; self.n_levels * self.dim];
        for level in 0..self.n_levels {
            for d in 0..self.dim {
                centroids[level * self.dim + d] = self.vmin[d] + (level as f32 + 0.5) / self.n_levels as f32 * self.vrange[d];
            }
        }
        centroids
    }

    fn score_quantized(&self, query: &[f32], indices: &[usize]) -> Vec<f32> {
        // Build lookup table: lut[level * dim + d] = query[d] * centroid[level, d]
        let mut lut = vec![0.0f32; self.n_levels * self.dim];
        for level in 0..self.n_levels {
            for d in 0..self.dim {
                lut[level * self.dim + d] = query[d] * self.centroids[level * self.dim + d];
            }
        }

        indices.iter().map(|&idx| {
            let mut score = 0.0f32;
            for d in 0..self.dim {
                let level = self.quant_codes[idx * self.dim + d] as usize;
                score += lut[level * self.dim + d];
            }
            score
        }).collect()
    }

    pub fn len(&self) -> usize {
        self.n_vectors
    }

    pub fn is_empty(&self) -> bool {
        self.n_vectors == 0
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_three_stage_basic() {
        let dim = 16;
        let mut index = ThreeStageIndex::new(dim, 50, 10, 4);

        let vectors: Vec<f32> = (0..dim * 50)
            .map(|i| ((i * 13 + 7) % 19) as f32 - 9.0)
            .collect();

        index.add(&vectors);
        assert_eq!(index.len(), 50);

        let query = &vectors[0..dim];
        let (scores, indices) = index.search(query, 5);
        assert!(!indices.is_empty());
        assert!(scores[0] > 0.9);
    }
}
